xdsl/transforms/csl_stencil_handle_async_flow.py

```
@@ -0,0 +1,233 @@
+    Loop var and iter_args are promoted to module-level csl vars.
+
+    Limitations:
+      * Loop can only yield its input iter_args (in any order)
+      * Loop bounds and step must be arith.constant values
+    """
+
+    counter: int
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: scf.For, rewriter: PatternRewriter, /):
+        if not self._is_inside_wrapper_outside_apply(op):
+            return
+
+        parent_func = op.parent_op()
+        while parent_func is not None and not isinstance(parent_func, csl.FuncOp):
+            parent_func = parent_func.parent_op()
+        if not parent_func:
+            return
+
+        # limitation: can yield iter_args in any order, but they cannot be modified in the loop body
+        terminator = op.body.block.last_op
+        assert isinstance(terminator, scf.Yield)
+        assert all(
+            arg in op.body.block.args for arg in terminator.arguments
+        ), "Can only yield unmodified iter_args (in any order)"
+
+        # limitation: currently only loops built from arith.constant are supported
+        assert isinstance(op.lb, OpResult)
+        assert isinstance(op.ub, OpResult)
+        assert isinstance(op.step, OpResult)
+        assert isinstance(op.lb.op, arith.Constant)
+        assert isinstance(op.ub.op, arith.Constant)
+        assert isinstance(op.step.op, arith.Constant)
+        assert isa(op.lb.op.value, IntegerAttr[IndexType])
+        assert isa(op.ub.op.value, IntegerAttr[IndexType])
+        assert isa(op.step.op.value, IntegerAttr[IndexType])
+
+        # limitation: all iter_args must be memrefs (stencil buffers) and have the same data type
+        assert isa(op.iter_args[0].type, MemRefType[csl.ZerosOp.T])
+        element_type = op.iter_args[0].type.get_element_type()
+        assert all(
+            isa(a.type, MemRefType[csl.ZerosOp.T])
+            and element_type == a.type.get_element_type()
+            for a in op.iter_args
+        )
```

Nit: I'm generally not a fan of asserts in rewrite patterns, but I can see the appeal here. When the pass just aborts you may end up with silent failures, which are a pain to debug... I guess this shows that xDSL is lacking some sort of warning system for not-applied passes...

---

I think in MLIR there are separate match and apply functions, which separates out code that checks if a pattern should be applied. We also need quite a few of these asserts to make pyright happy, and the new `make precommit` forces us to split `assert isinstance(op.lb, OpResult) and isinstance(op.lb.op, arith.Constant)` into separate asserts.